Initial version

ddragosd opened the pull request:
- sync with `rclone`.  Note that `rclone` [doesn't work](https://github.com/ncw/rclone/issues/85) yet with IAM Roles and this is why we're using AWS cli. 
- includes AWS CLI for syncing files form S3

ddragosd committed:
- initial prototype

ddragosd committed:
Docker builds a minimalistic image of 28MB

ddragosd committed:
added the main file which was missed from the initial commit

ddragosd committed:
updated README

ddragosd committed:
added aws-cli to the container for convenience

ddragosd committed:
if the gateway fails to reload b/c a configuration file has an error, don't exit the sync job as it may get fixed later

clehene commented:
:+1:

selfxp commented:
:+1:

adragomir commented on config-supervisor.go:
```diff
@@ -0,0 +1,127 @@
+	"runtime"
+	"runtime/pprof"
+	"time"
+	//    "net/http"
```
Remove please

adragomir commented on config-supervisor.go:
```diff
@@ -0,0 +1,127 @@
+)
+
+func syntaxError() {
+	fmt.Fprintf(os.Stderr, `Execute a sync command and watch a folder for changes.
```
Move to same line. Is this after go fmt ?

adragomir commented on sync/recursiveWatcher.go:
```diff
@@ -0,0 +1,126 @@
+package sync
```
the package file convention for go is lowercase ?

ddragosd replied:
yes, that's what I noticed

adragomir commented on sync/recursiveWatcher.go:
```diff
@@ -0,0 +1,126 @@
+			select {
+			case event := <-watcher.Events:
+				// create a file/directory
+				if event.Op&fsnotify.Create == fsnotify.Create {
```
Separate & with spaces

ddragosd replied:
@adragomir I came to learn this is a result of running  `gofmt -e -w ./`
or `make format`

ddragosd replied:
For Ref: https://golang.org/cmd/gofmt/

adragomir commented on sync/recursiveWatcher.go:
```diff
@@ -0,0 +1,126 @@
+		if info.IsDir() {
+			name := info.Name()
+			// skip folders that begin with a dot
+			if shouldIgnoreFile(name) && name != "." && name != ".." {
```
move comparison to shouldIgnoreFile ?

adragomir commented:
@ddragosd :metal: after the comments. I'll create an issue to use sync from the go AWS SDK, I really don't like executing aws s3 sync from the command line.

ddragosd commented:
@adragomir I've also started initially at integrating a library as opposed to executing a command line. I've picked [rclone](https://gowalker.org/github.com/ncw/rclone) as it comes with broader support, it's not just for AWS. 

There are 2 problems with `rclone`:
1. I couldn't integrate it as a lib - hence the idea of executing a command line
2. It doesn't work with IAM Roles :smile:  - which is a big miss for people using AWS. If the 1st point would have worked we could have fixed the IAM roles.

So at the moment I don't have a solution to support multiple clouds and integrate it as a lib, but I agree with you that it would be a nicer implementation.

ddragosd committed:
implemented code reviews and added a simple unit test

ddragosd committed:
added another unit test to check that sync cmd executes

ddragosd committed:
health-check prints more information about the sync activity :
```
{
    status: "OK",
    lastSync: "2016-02-24T23:10:36.148512004-08:00",
    lastChangeDetected: "2016-02-24T23:07:36.036041758-08:00",
    lastReload: "2016-02-24T23:07:41.041648492-08:00"
}
```

ddragosd commented:
@selfxp @clehene @adragomir I've added unit tests and along with them I've enriched the `health-check` API.  This API may be exposed also through the API Gateway's `health-check` and even take action in the Gateway if the `lastSync` is not as expected ( i.e < 10s ), and return a `5xx` error, causing the container to restart. 

(updated) Sample output for the health-check API:

```
{
    status: "OK",
    lastSync: "2016-02-24T23:10:36.148512004-08:00",
    lastChangeDetected: "2016-02-24T23:07:36.036041758-08:00",
    lastReload: "2016-02-24T23:07:41.041648492-08:00"
}
```

Let me know what you think.

ddragosd committed:
`go fmt` updated source files

ddragosd committed:
updated Dockerfile to execute the unit tests during build

selfxp commented:
👍
